agones/src/lib.rs

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+use std::{
+    collections::BTreeMap,
+    env,
+    time::{SystemTime, UNIX_EPOCH},
+};
+
+use k8s_openapi::{api::core::v1::Namespace, apimachinery::pkg::apis::meta::v1::ObjectMeta};
+use kube::{
+    api::{DeleteParams, ListParams, PostParams},
+    Api, ResourceExt,
+};
+use tokio::sync::OnceCell;
+
+mod pod;
+
+#[allow(dead_code)]
+static CLIENT: OnceCell<Client> = OnceCell::const_new();
+#[allow(dead_code)]
+const QUILKIN_IMAGE: &str = "QUILKIN_IMAGE";
+
+pub struct Client {
+    /// The Kubernetes client
+    pub kubernetes: kube::Client,
+    /// The namespace the tests will happen in
+    pub namespace: String,
+    /// The name and tag of the Quilkin image being tested
+    pub quilkin_image: String,
+}
+
+impl Client {
+    /// Thread safe way to create a Client once and only once across multiple tests.
+    /// Executes the setup required:
+    /// * Creates a test namespace for this test
+    /// * Removes previous test namespaces
+    /// * Retrieves the QUILKIN_IMAGE to test from env vars, and defaults it if not available.
+    pub async fn new() -> &'static Client {
+        CLIENT
+            .get_or_init(|| async {
+                let client = kube::Client::try_default()
+                    .await
+                    .expect("Kubernetes client to be created");
+
+                Client {
+                    kubernetes: client.clone(),
+                    namespace: setup_namespace(client).await,
+                    quilkin_image: env::var_os(QUILKIN_IMAGE)
+                        .unwrap_or_else(|| "us-docker.pkg.dev/quilkin/release/quilkin:0.3.0".into())
+                        .into_string()
+                        .unwrap(),
```

Comment:
It would be much nicer if by default this built a local image, rather than use a previous release. In my experience that kind of default can give me a false sense of security as you'll occasionally forget to set the correct environment variable and test the wrong code. It's also a hassle to have to remember to set it compared to having `cargo test --all` test everything correctly out of the box without setup from the developer.

I don't think it's an unreasonable assumption to assume here that if someone hasn't explicitly set `QUILKIN_IMAGE`, then we're being called locally from cargo, so you'd be able to call `make build-image` to generate a new build. If Quilkin's build image was a multi-stage dockerfile you could probably use a docker client like [`bollard`](https://docs.rs/bollard/latest/bollard/index.html) to build and get it without make.

Reply:
> It would be much nicer if by default this built a local image, rather than use a previous release

Hmnn. Interesting idea. That's where I figured we would end up eventually for sure, this was just a first step.

Could likely also it via a Make target to do the whole thing and/or reuse the `Make` tooling via `std::process::Command`, or use something like bollard as suggested (although that would introduce a lot of duplication since we are generating executable artifacts and a runnable images, and we would need to keep the builder step and the builder image still in sync. 

There are a lot of unanswered questions here though if it was automatically built, which is why I ended up here, at least for a first pass.

1. Which registry is the end user using? I can't hardcode the registry into the tag that gets built.
2. Do you need to push to a registry, or push to a minikube instance? -- without pushing the local image somewhere, it can't be run. on a cluster.

> In my experience that kind of default can give me a false sense of security as you'll occasionally forget to set the correct environment variable and test the wrong code. 

🤔 I went back and forth on this one too, I originally thought of making the env var required, and have the system panic if it wasn't set -- so there was no possible way to skip it. For this pass, I'd recommend that we make it required, and then look at how we can make the development iteration loop smoother? (build+push+test in one go)

> I don't think it's an unreasonable assumption to assume here that if someone hasn't explicitly set QUILKIN_IMAGE, then we're being called locally from cargo, so you'd be able to call make build-image to generate a new build.

Definitely don't disagree! The other way to solve would be to extend `make test-agones` to do very much the same thing -- that way you could set environment or make variables to specify things like registry and where to push images -- i.e. could do a `make test-minikube-agones` that works a little differently for minikube -- but this is all discuss-able with different tradeoffs.

But we'll need some configuration options somewhere for different container and cluster environments.